sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,305 @@
+
+ - after spec constant enumeration (symbolic -\> int ID translation), the SC pass
+ will handle the `__sycl_getCompositeSpecConstantValue`. Knowning the composite
+ specialization constant's type (`%struct.POD`), the pass will traverse its leaf
+ fields and generate 5 "primitive" spec constants using already existing SPIR-V intrinsic:
+
+```cpp
+%gold_POD_a0x = call i32 __spirv_SpecConstant(i32 10, i32 0)
+%gold_POD_a0y = call float __spirv_SpecConstant(i32 11, float 0)
+%gold_POD_a1x = call i32 __spirv_SpecConstant(i32 12, i32 0)
+%gold_POD_a1y = call float __spirv_SpecConstant(i32 13, float 0)
+%gold_POD_b = call i32 __spirv_SpecConstant(i32 14, i32 0)
+```
+
+And 1 "composite"
+
+```cpp
+  %gold_POD = call %struct.POD __spirvCompositeSpecConstant<POD mangling>(10, 11, 12, 13, 14, 15)
+```
+
+where `__spirvCompositeSpecConstant<type mangling>` is a new SPIR-V intrinsic which
+ represents creation of a composite specialization constant. Its arguments are spec
+ constant IDs corresponding to the leaf fields of the POD type of the constant.
+ ID is not needed, as runtime will never use it - it will use IDs of the leaves instead.
+ Yet SPIR-V does not allow IDs for composite spec constants.
+
+##### The post-link tool changes
+
+For composite specialization constants the post link tool will additionally
+generate linearized list of \<leaf spec ID,type,offset,size\> tuples (descriptors),
+where each tuple describes a leaf field, and store it together with the
+existing meta-information associated with the specialization constants and
+passed to the runtime. Also, for a composite specialization constant the will be
+no ID map entry within the meta information, and the composite constant will
+referenced by its symbolic ID. For example:
+
+```cpp
+MyConst_mangled [10,int,0,4],[10,float,4,4],[10,int,8,4],[10,float,12,4],[10,int,16,4]
+```
+
+#### LLVMIR-\>SPIR-V translator
+
+The translator aims to create the following code (pseudo-code)
+
+```cpp
+%gold_POD_a0x = OpSpecConstant(0)    [SpecId = 10]
+%gold_POD_a0y = OpSpecConstant(0.0f) [SpecId = 11]
+%gold_POD_a1x = OpSpecConstant(0)    [SpecId = 12]
+%gold_POD_a1y = OpSpecConstant(0.0f) [SpecId = 13]
+%gold_POD_b   = OpSpecConstant(0)    [SpecId = 14]
+
+%gold_POD_a0 = OpSpecConstantComposite(
+  %gold_POD_a0x // gold.a[0].x
+  %gold_POD_a0y // gold.a[0].y
+)
+
+%gold_POD_a1 = OpSpecConstantComposite(
+  %gold_POD_a1x // gold.a[1].x
+  %gold_POD_a1y // gold.a[1].y
+)
+
+%gold_POD = OpSpecConstantComposite(
+  %gold_POD_a0,
+  %gold_POD_a1,
+  %gold_POD_b   // gold.b
+}
+```
+
+ - `OpSpecConstant` operations will be created using already existing mechanism for
+ the primitive spec constants.
+ - Then the translator will handle `__spirvCompositeSpecConstant*` intrinsic. It will
+ recursively traverse the spec constant type structure in parallel with the argument
+ list - which is a list of primitive spec constant operation IDs (not their SpecIds!).
```

Comment:
I'm a bit confused. What is the argument list? Is it the argument list of `__spirvCompositeSpecConstant*`? But the example above says that it is a list of SpecId indeed. Having SpecIds here looks logical to me.

Reply:
good catch - will change. It should be SpecIDs of course

Reply:
It is the SPIRV OpSpecConstantComposite operation, where arguments (operands) are spec constant result IDs rather than SpecIDs